src/ansi.rs

```diff
@@ -0,0 +1,1912 @@
+        }
+
+        if params.is_empty() || params[0].is_empty() {
+            return;
+        }
+
+        match params[0] {
+            // Set window title.
+            b"0" | b"2" => {
+                if params.len() >= 2 {
+                    let title = params[1..]
+                        .iter()
+                        .flat_map(|x| str::from_utf8(x))
+                        .collect::<Vec<&str>>()
+                        .join(";")
+                        .trim()
+                        .to_owned();
+                    self.handler.set_title(Some(title));
+                    return;
+                }
+                unhandled(params);
+            },
+
+            // Set color index.
+            b"4" => {
+                if params.len() <= 1 || params.len() % 2 == 0 {
+                    unhandled(params);
+                    return;
+                }
+
+                for chunk in params[1..].chunks(2) {
+                    let index = match parse_number(chunk[0]) {
+                        Some(index) => index,
+                        None => {
+                            unhandled(params);
+                            continue;
+                        },
+                    };
+
+                    if let Some(c) = xparse_color(chunk[1]) {
+                        self.handler.set_color(index as usize, c);
+                    } else if chunk[1] == b"?" {
+                        let prefix = alloc::format!("4;{index}");
+                        self.handler.dynamic_color_sequence(prefix, index as usize, terminator);
+                    } else {
+                        unhandled(params);
+                    }
+                }
+            },
+
+            // Hyperlink.
+            b"8" if params.len() > 2 => {
+                let link_params = params[1];
+
+                // NOTE: The escape sequence is of form 'OSC 8 ; params ; URI ST', where
+                // URI is URL-encoded. However `;` is a special character and might be
+                // passed as is, thus we need to rebuild the URI.
+                let mut uri = str::from_utf8(params[2]).unwrap_or_default().to_string();
+                for param in params[3..].iter() {
+                    uri.push(';');
+                    uri.push_str(str::from_utf8(param).unwrap_or_default());
+                }
+
+                // The OSC 8 escape sequence must be stopped when getting an empty `uri`.
+                if uri.is_empty() {
+                    self.handler.set_hyperlink(None);
+                    return;
+                }
+
+                // Link parameters are in format of `key1=value1:key2=value2`. Currently only key
+                // `id` is defined.
+                let id = link_params
+                    .split(|&b| b == b':')
+                    .find_map(|kv| kv.strip_prefix(b"id="))
+                    .and_then(|kv| str::from_utf8(kv).ok().map(|e| e.to_owned()));
```

Comment:
I'd assume .cloned should work here?

Reply:
`cloned()` will not work because it will return a `str` (not `Sized` so would'nt compile anyways) while we are trying to get a `String`.